From 0613345efb8cd66e7be23bbacbc42bebad345d2c Mon Sep 17 00:00:00 2001
From: Adithya R <gh0strider.2k18.reborn@gmail.com>
Date: Tue, 20 Jun 2023 22:58:41 +0530
Subject: [PATCH] SystemUI: Introduce HD & wifi calling statusbar icons

HD icon shows for VoLTE/VoNR, and wifi calling for VoWiFi.
Based on CLO's volte/vowifi icon implementation, reworked manually.
Drawables taken from Nothing OS and manually tweaked.

[@techyminati - VollaOS - Adapt code for VollaOS 13)

Squashed:

Author: Adithya R <gh0strider.2k18.reborn@gmail.com>
Date:   Mon Jul 3 15:07:44 2023 +0530

    SystemUI: Fix HD calling icon for RTL layout

    The mobile signal icon mirrors itself in RTL mode, so the HD icon
    needs to be moved from top left to top right.

    Before: https://i.imgur.com/tUGTtC2.jpg
    After: https://i.imgur.com/0VWMOrC.jpg

    Change-Id: I20a6c6205c20e189247023b0f4429a27871765dc

Co-authored-by: Weijie Wang <quic_weijiew@quicinc.com>
Change-Id: I0190394690e9b38e0575ffb099be8a5ed2ae9d90
Signed-off-by: Mohammad Hasan Keramat J <ikeramat@protonmail.com>
Signed-off-by: Adithya R <gh0strider.2k18.reborn@gmail.com>
---
 .../SettingsLib/res/values/volla_config.xml   |  29 ++++
 .../settingslib/graph/SignalDrawable.java     |  35 ++++-
 .../status_bar_mobile_signal_group_inner.xml  |  23 ++-
 .../SystemUI/res/drawable/ic_hd_calling.xml   |  13 ++
 .../res/drawable/ic_statusbar_hd_calling.xml  |  17 +++
 .../res/drawable/ic_statusbar_roaming.xml     |  30 ++++
 .../res/drawable/ic_statusbar_vowifi.xml      |  14 ++
 packages/SystemUI/res/drawable/ic_vowifi.xml  |  13 ++
 packages/SystemUI/res/values/cm_strings.xml   |  49 ++++++
 .../SystemUI/res/values/lineage_dimens.xml    |   7 +
 .../SystemUI/res/xml/status_bar_prefs.xml     |  15 ++
 .../statusbar/StatusBarMobileView.java        |  31 ++--
 .../connectivity/MobileSignalController.java  | 139 +++++++++++++++++-
 .../statusbar/connectivity/MobileState.kt     |  22 +++
 .../statusbar/connectivity/SignalCallback.kt  |   2 +
 .../connectivity/WifiSignalController.java    |   2 +-
 .../phone/StatusBarSignalPolicy.java          |  31 ++--
 17 files changed, 428 insertions(+), 44 deletions(-)
 create mode 100644 packages/SettingsLib/res/values/volla_config.xml
 create mode 100644 packages/SystemUI/res/drawable/ic_hd_calling.xml
 create mode 100644 packages/SystemUI/res/drawable/ic_statusbar_hd_calling.xml
 create mode 100644 packages/SystemUI/res/drawable/ic_statusbar_roaming.xml
 create mode 100644 packages/SystemUI/res/drawable/ic_statusbar_vowifi.xml
 create mode 100644 packages/SystemUI/res/drawable/ic_vowifi.xml

diff --git a/packages/SettingsLib/res/values/volla_config.xml b/packages/SettingsLib/res/values/volla_config.xml
new file mode 100644
index 0000000000000..66ba90a1ad4a8
--- /dev/null
+++ b/packages/SettingsLib/res/values/volla_config.xml
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+     Copyright (C) 2021 The LibreMobileOS Foundation
+     Copyright (C) 2024 Volla Systeme GmbH
+
+     Licensed under the Apache License, Version 2.0 (the "License");
+     you may not use this file except in compliance with the License.
+     You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+     Unless required by applicable law or agreed to in writing, software
+     distributed under the License is distributed on an "AS IS" BASIS,
+     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+     See the License for the specific language governing permissions and
+     limitations under the License.
+-->
+<resources>
+
+    <!-- Roaming symbol (R) path for SignalDrawable as defined on a 24x24 canvas. -->
+    <string name="config_signalRoamingPath" translatable="false">
+        M 19.436 19.688 L 18.152 19.688 L 18.152 22 L 17 22 L 17 16 L 19.332 16 C 20.097 16 20.687 16.155 21.103 16.466 C 21.519 16.776 21.727 17.225 21.727 17.813 C 21.727 18.214 21.619 18.551 21.403 18.823 C 21.191 19.092 20.893 19.299 20.511 19.445 L 22 21.946 L 22 22 L 20.766 22 L 19.436 19.688 Z M 18.152 18.852 L 19.336 18.852 C 19.725 18.852 20.028 18.764 20.247 18.588 C 20.465 18.409 20.575 18.166 20.575 17.859 C 20.575 17.537 20.473 17.288 20.27 17.113 C 20.069 16.937 19.769 16.846 19.368 16.841 L 18.152 16.841 L 18.152 18.852 Z
+    </string>
+    <!-- config_roamingCutout{Height,Width}Fraction define fraction of the 24x24 canvas that
+         should be cut out to display config_signalRoamingPath. -->
+    <item name="config_roamingCutoutWidthFraction" format="float" type="dimen">8</item>
+    <item name="config_roamingCutoutHeightFraction" format="float" type="dimen">9</item>
+
+</resources>
diff --git a/packages/SettingsLib/src/com/android/settingslib/graph/SignalDrawable.java b/packages/SettingsLib/src/com/android/settingslib/graph/SignalDrawable.java
index 4d0804e1bbb78..e40222539b8b7 100644
--- a/packages/SettingsLib/src/com/android/settingslib/graph/SignalDrawable.java
+++ b/packages/SettingsLib/src/com/android/settingslib/graph/SignalDrawable.java
@@ -62,7 +62,8 @@ public class SignalDrawable extends DrawableWrapper {
     private static final int STATE_SHIFT = 16;
     private static final int STATE_MASK = 0xff << STATE_SHIFT;
     private static final int STATE_CUT = 2;
-    private static final int STATE_CARRIER_CHANGE = 3;
+    private static final int STATE_CUT_R = 3;
+    private static final int STATE_CARRIER_CHANGE = 4;
 
     private static final long DOT_DELAY = 1000;
 
@@ -73,11 +74,15 @@ public class SignalDrawable extends DrawableWrapper {
     private final Path mCutoutPath = new Path();
     private final Path mForegroundPath = new Path();
     private final Path mAttributionPath = new Path();
+    private final Path mRoamingPath = new Path();
     private final Matrix mAttributionScaleMatrix = new Matrix();
     private final Path mScaledAttributionPath = new Path();
+    private final Path mScaledRoamingPath = new Path();
     private final Handler mHandler;
     private final float mCutoutWidthFraction;
     private final float mCutoutHeightFraction;
+    private final float mRCutoutWidthFraction;
+    private final float mRCutoutHeightFraction;
     private float mDarkIntensity = -1;
     private final int mIntrinsicSize;
     private boolean mAnimating;
@@ -87,12 +92,19 @@ public class SignalDrawable extends DrawableWrapper {
         super(context.getDrawable(com.android.internal.R.drawable.ic_signal_cellular));
         final String attributionPathString = context.getString(
                 com.android.internal.R.string.config_signalAttributionPath);
+        final String roamingPathString = context.getString(
+                R.string.config_signalRoamingPath);
         mAttributionPath.set(PathParser.createPathFromPathData(attributionPathString));
+        mRoamingPath.set(PathParser.createPathFromPathData(roamingPathString));
         updateScaledAttributionPath();
         mCutoutWidthFraction = context.getResources().getFloat(
                 com.android.internal.R.dimen.config_signalCutoutWidthFraction);
         mCutoutHeightFraction = context.getResources().getFloat(
                 com.android.internal.R.dimen.config_signalCutoutHeightFraction);
+        mRCutoutWidthFraction = context.getResources().getFloat(
+                R.dimen.config_roamingCutoutWidthFraction);
+        mRCutoutHeightFraction = context.getResources().getFloat(
+                R.dimen.config_roamingCutoutHeightFraction);
         mDarkModeFillColor = Utils.getColorStateListDefaultColor(context,
                 R.color.dark_mode_icon_color_single_tone);
         mLightModeFillColor = Utils.getColorStateListDefaultColor(context,
@@ -112,6 +124,7 @@ public class SignalDrawable extends DrawableWrapper {
                     getBounds().width() / VIEWPORT, getBounds().height() / VIEWPORT);
         }
         mAttributionPath.transform(mAttributionScaleMatrix, mScaledAttributionPath);
+        mRoamingPath.transform(mAttributionScaleMatrix, mScaledRoamingPath);
     }
 
     @Override
@@ -213,16 +226,20 @@ public class SignalDrawable extends DrawableWrapper {
             drawDotAndPadding(x - dotSpacing * 2, y, dotPadding, dotSize, 0);
             canvas.drawPath(mCutoutPath, mTransparentPaint);
             canvas.drawPath(mForegroundPath, mForegroundPaint);
-        } else if (isInState(STATE_CUT)) {
-            float cutX = (mCutoutWidthFraction * width / VIEWPORT);
-            float cutY = (mCutoutHeightFraction * height / VIEWPORT);
+        } else if (isInState(STATE_CUT) || isInState(STATE_CUT_R)) {
+            boolean isRoaming = isInState(STATE_CUT_R);
+            float cutWidth = isRoaming ? mRCutoutWidthFraction : mCutoutWidthFraction;
+            float cutHeight = isRoaming ? mRCutoutHeightFraction : mCutoutHeightFraction;
+            float cutX = (cutWidth * width / VIEWPORT);
+            float cutY = (cutHeight * height / VIEWPORT);
             mCutoutPath.moveTo(width, height);
             mCutoutPath.rLineTo(-cutX, 0);
             mCutoutPath.rLineTo(0, -cutY);
             mCutoutPath.rLineTo(cutX, 0);
             mCutoutPath.rLineTo(0, cutY);
             canvas.drawPath(mCutoutPath, mTransparentPaint);
-            canvas.drawPath(mScaledAttributionPath, mForegroundPaint);
+            canvas.drawPath(isRoaming ? mScaledRoamingPath : mScaledAttributionPath,
+                    mForegroundPaint);
         }
         if (isRtl) {
             canvas.restore();
@@ -284,12 +301,16 @@ public class SignalDrawable extends DrawableWrapper {
         return (fullState & STATE_MASK) >> STATE_SHIFT;
     }
 
-    public static int getState(int level, int numLevels, boolean cutOut) {
-        return ((cutOut ? STATE_CUT : 0) << STATE_SHIFT)
+    public static int getState(int level, int numLevels, boolean cutOut, boolean roaming) {
+        return ((cutOut ? STATE_CUT : roaming ? STATE_CUT_R : 0) << STATE_SHIFT)
                 | (numLevels << NUM_LEVEL_SHIFT)
                 | level;
     }
 
+    public static int getState(int level, int numLevels, boolean cutOut) {
+        return getState(level, numLevels, cutOut, false);
+    }
+
     /** Returns the state representing empty mobile signal with the given number of levels. */
     public static int getEmptyState(int numLevels) {
         return getState(0, numLevels, true);
diff --git a/packages/SystemUI/res-keyguard/layout/status_bar_mobile_signal_group_inner.xml b/packages/SystemUI/res-keyguard/layout/status_bar_mobile_signal_group_inner.xml
index 29832a081612d..1c30fe7910ac9 100644
--- a/packages/SystemUI/res-keyguard/layout/status_bar_mobile_signal_group_inner.xml
+++ b/packages/SystemUI/res-keyguard/layout/status_bar_mobile_signal_group_inner.xml
@@ -50,6 +50,12 @@
                 android:visibility="gone"
                 />
         </FrameLayout>
+        <Space
+            android:id="@+id/mobile_type_space"
+            android:layout_height="match_parent"
+            android:layout_width="1.5dp"
+            android:visibility="gone"
+        />
         <ImageView
             android:id="@+id/mobile_type"
             android:layout_height="wrap_content"
@@ -59,9 +65,9 @@
             android:paddingEnd="1dp"
             android:visibility="gone" />
         <Space
-            android:id="@+id/mobile_roaming_space"
+            android:id="@+id/mobile_hd_space"
             android:layout_height="match_parent"
-            android:layout_width="@dimen/roaming_icon_start_padding"
+            android:layout_width="2.5dp"
             android:visibility="gone"
             />
         <FrameLayout
@@ -75,11 +81,14 @@
                 systemui:hasOverlappingRendering="false"
                 />
             <ImageView
-                android:id="@+id/mobile_roaming"
-                android:layout_width="wrap_content"
-                android:layout_height="wrap_content"
-                android:src="@drawable/stat_sys_roaming"
-                android:contentDescription="@string/data_connection_roaming"
+                android:id="@+id/mobile_hd"
+                android:layout_width="12dp"
+                android:layout_height="8dp"
+                android:layout_marginTop="1dp"
+                android:layout_gravity="start|top"
+                android:src="@drawable/ic_hd_calling"
+                android:adjustViewBounds="true"
+                android:gravity="start|top"
                 android:visibility="gone" />
         </FrameLayout>
         <ImageView
diff --git a/packages/SystemUI/res/drawable/ic_hd_calling.xml b/packages/SystemUI/res/drawable/ic_hd_calling.xml
new file mode 100644
index 0000000000000..0a3bddf6e8ffc
--- /dev/null
+++ b/packages/SystemUI/res/drawable/ic_hd_calling.xml
@@ -0,0 +1,13 @@
+<vector
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="12dp"
+    android:height="8dp"
+    android:viewportWidth="12"
+    android:viewportHeight="8">
+    <path
+        android:pathData="M 5.402 7.6 L 4.225 7.6 L 4.225 4.401 L 1.182 4.401 L 1.182 7.6 L 0 7.6 L 0 0.4 L 1.182 0.4 L 1.182 3.396 L 4.225 3.396 L 4.225 0.4 L 5.402 0.4 L 5.402 7.6 Z"
+        android:fillColor="#000000"/>
+    <path
+        android:pathData="M 6.924 7.6 L 6.924 0.4 L 8.934 0.4 C 9.535 0.4 10.068 0.542 10.532 0.825 C 10.999 1.109 11.362 1.511 11.617 2.032 C 11.872 2.553 12 3.149 12 3.822 L 12 4.182 C 12 4.865 11.87 5.465 11.612 5.982 C 11.357 6.501 10.991 6.899 10.513 7.18 C 10.04 7.46 9.497 7.6 8.882 7.6 L 6.924 7.6 Z M 8.107 1.409 L 8.107 6.602 L 8.878 6.602 C 9.498 6.602 9.973 6.396 10.303 5.988 C 10.637 5.576 10.807 4.985 10.813 4.217 L 10.813 3.818 C 10.813 3.035 10.652 2.439 10.332 2.027 C 10.01 1.614 9.545 1.409 8.934 1.409 L 8.107 1.409 Z"
+        android:fillColor="#000000"/>
+</vector>
diff --git a/packages/SystemUI/res/drawable/ic_statusbar_hd_calling.xml b/packages/SystemUI/res/drawable/ic_statusbar_hd_calling.xml
new file mode 100644
index 0000000000000..c97a9d45efa71
--- /dev/null
+++ b/packages/SystemUI/res/drawable/ic_statusbar_hd_calling.xml
@@ -0,0 +1,17 @@
+<vector
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="24dp"
+    android:height="24dp"
+    android:viewportWidth="24"
+    android:viewportHeight="24"
+    android:tint="?android:attr/colorControlNormal">
+    <path
+        android:pathData="M 22 2 L 2 22 L 22 22 L 22 2 L 22 2 Z"
+        android:fillColor="#000000"/>
+    <path
+        android:pathData="M 5.402 8.2 L 4.225 8.2 L 4.225 5.001 L 1.182 5.001 L 1.182 8.2 L 0 8.2 L 0 1 L 1.182 1 L 1.182 3.996 L 4.225 3.996 L 4.225 1 L 5.402 1 L 5.402 8.2 Z"
+        android:fillColor="#000000"/>
+    <path
+        android:pathData="M 6.924 8.2 L 6.924 1 L 8.934 1 C 9.535 1 10.068 1.142 10.532 1.425 C 10.999 1.709 11.362 2.111 11.617 2.632 C 11.872 3.153 12 3.749 12 4.422 L 12 4.782 C 12 5.465 11.87 6.065 11.612 6.582 C 11.357 7.101 10.991 7.499 10.513 7.78 C 10.04 8.06 9.497 8.2 8.882 8.2 L 6.924 8.2 Z M 8.107 2.009 L 8.107 7.202 L 8.878 7.202 C 9.498 7.202 9.973 6.996 10.303 6.588 C 10.637 6.176 10.807 5.585 10.813 4.817 L 10.813 4.418 C 10.813 3.635 10.652 3.039 10.332 2.627 C 10.01 2.214 9.545 2.009 8.934 2.009 L 8.107 2.009 Z"
+        android:fillColor="#000000"/>
+</vector>
diff --git a/packages/SystemUI/res/drawable/ic_statusbar_roaming.xml b/packages/SystemUI/res/drawable/ic_statusbar_roaming.xml
new file mode 100644
index 0000000000000..25e0303d16f70
--- /dev/null
+++ b/packages/SystemUI/res/drawable/ic_statusbar_roaming.xml
@@ -0,0 +1,30 @@
+<!--
+     Copyright (C) 2021 The Android Open Source Project
+
+     Licensed under the Apache License, Version 2.0 (the "License");
+     you may not use this file except in compliance with the License.
+     You may obtain a copy of the License at
+
+          http://www.apache.org/licenses/LICENSE-2.0
+
+     Unless required by applicable law or agreed to in writing, software
+     distributed under the License is distributed on an "AS IS" BASIS,
+     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+     See the License for the specific language governing permissions and
+     limitations under the License.
+-->
+<vector
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="24dp"
+    android:height="24dp"
+    android:viewportWidth="24"
+    android:viewportHeight="24"
+    android:tint="?android:attr/colorControlNormal">
+    <path
+        android:pathData="M 22 2 L 2 22 L 22 22 L 22 2 L 22 2 Z M 16 15 L 22 15 L 22 22 L 16 22 L 16 15"
+        android:fillColor="#000000"
+        android:fillType="evenOdd"/>
+    <path
+        android:pathData="M 19.436 19.688 L 18.152 19.688 L 18.152 22 L 17 22 L 17 16 L 19.332 16 C 20.097 16 20.687 16.155 21.103 16.466 C 21.519 16.776 21.727 17.225 21.727 17.813 C 21.727 18.214 21.619 18.551 21.403 18.823 C 21.191 19.092 20.893 19.299 20.511 19.445 L 22 21.946 L 22 22 L 20.766 22 L 19.436 19.688 Z M 18.152 18.852 L 19.336 18.852 C 19.725 18.852 20.028 18.764 20.247 18.588 C 20.465 18.409 20.575 18.166 20.575 17.859 C 20.575 17.537 20.473 17.288 20.27 17.113 C 20.069 16.937 19.769 16.846 19.368 16.841 L 18.152 16.841 L 18.152 18.852 Z"
+        android:fillColor="#000000"/>
+</vector>
diff --git a/packages/SystemUI/res/drawable/ic_statusbar_vowifi.xml b/packages/SystemUI/res/drawable/ic_statusbar_vowifi.xml
new file mode 100644
index 0000000000000..35191a2d9eb3c
--- /dev/null
+++ b/packages/SystemUI/res/drawable/ic_statusbar_vowifi.xml
@@ -0,0 +1,14 @@
+<vector
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="24dp"
+    android:height="24dp"
+    android:viewportWidth="24"
+    android:viewportHeight="24"
+    android:tint="?android:attr/colorControlNormal">
+    <path
+        android:pathData="M 21.996 4.95 C 21.787 4.779 19.666 2.999 16.497 2.999 C 13.318 2.999 11.208 4.779 10.999 4.95 L 16.497 11.998 L 21.996 4.95 Z"
+        android:fillColor="#000000"/>
+    <path
+        android:pathData="M 19.997 15.507 C 18.756 15.507 17.548 15.308 16.427 14.938 C 16.077 14.818 15.678 14.907 15.407 15.178 L 13.208 17.377 C 10.378 15.928 8.059 13.618 6.62 10.788 L 8.819 8.589 C 9.098 8.309 9.178 7.919 9.069 7.569 C 8.699 6.45 8.498 5.25 8.498 3.999 C 8.498 3.45 8.049 2.999 7.499 2.999 L 3.999 2.999 C 3.45 2.999 2.999 3.45 2.999 3.999 C 2.999 13.388 10.608 20.997 19.997 20.997 C 20.547 20.997 20.997 20.547 20.997 19.997 L 20.997 16.508 C 20.997 15.957 20.547 15.507 19.997 15.507 Z"
+        android:fillColor="#000000"/>
+</vector>
diff --git a/packages/SystemUI/res/drawable/ic_vowifi.xml b/packages/SystemUI/res/drawable/ic_vowifi.xml
new file mode 100644
index 0000000000000..535cb6962d702
--- /dev/null
+++ b/packages/SystemUI/res/drawable/ic_vowifi.xml
@@ -0,0 +1,13 @@
+<vector
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="14dp"
+    android:height="14dp"
+    android:viewportWidth="14"
+    android:viewportHeight="14">
+    <path
+        android:pathData="M 12.833 2.888 C 12.711 2.788 11.474 1.75 9.625 1.75 C 7.77 1.75 6.539 2.788 6.417 2.888 L 9.625 7 L 12.833 2.888 Z"
+        android:fillColor="#000000"/>
+    <path
+        android:pathData="M 11.667 9.047 C 10.943 9.047 10.238 8.931 9.584 8.715 C 9.38 8.645 9.147 8.697 8.989 8.855 L 7.706 10.138 C 6.055 9.293 4.702 7.945 3.862 6.294 L 5.145 5.011 C 5.308 4.848 5.355 4.62 5.291 4.416 C 5.075 3.763 4.958 3.063 4.958 2.333 C 4.958 2.013 4.696 1.75 4.375 1.75 L 2.333 1.75 C 2.013 1.75 1.75 2.013 1.75 2.333 C 1.75 7.811 6.189 12.25 11.667 12.25 C 11.988 12.25 12.25 11.988 12.25 11.667 L 12.25 9.631 C 12.25 9.31 11.988 9.047 11.667 9.047 Z"
+        android:fillColor="#000000"/>
+</vector>
diff --git a/packages/SystemUI/res/values/cm_strings.xml b/packages/SystemUI/res/values/cm_strings.xml
index 43603f08061fb..1e84f63c55d58 100644
--- a/packages/SystemUI/res/values/cm_strings.xml
+++ b/packages/SystemUI/res/values/cm_strings.xml
@@ -120,4 +120,53 @@
     <string name="screenshot_delete_label">Delete</string>
     <!-- Content description indicating that tapping the element will allow deleting the screenshot [CHAR LIMIT=NONE] -->
     <string name="screenshot_delete_description">Delete screenshot</string>
+    <!-- Navbar -->
+    <string name="systemui_tuner_navbar_title">Layout</string>
+    <string name="navigation_bar_format_title">Navbar format</string>
+    <string name="preview">Preview</string>
+    <string name="keycode">Keycode</string>
+    <string name="no_home_title">No home button found</string>
+    <string name="no_home_message">A home button is required to be able to navigate this device. Please add a home button before saving.</string>
+    <string name="clipboard_description">The Clipboard allows items to be dragged directly to the clipboard. Items can also be dragged directly out of the clipboard when present.</string>
+    <string name="keycode_description">Keycode buttons allow keyboard keys to be added to the Navigation Bar. When pressed they emulate the selected keyboard key. First the key must be selected for the button, followed by an image to be shown on the button.</string>
+    <string name="start">Start</string>
+    <string name="center">Center</string>
+    <string name="end">End</string>
+    <string name="space">Spacer</string>
+    <string name="contextual">Contextual</string>
+    <string name="menu_ime">Menu / Keyboard Switcher</string>
+    <string name="ime_switcher">Keyboard Switcher</string>
+    <string name="select_button">Select button to add</string>
+    <string name="add_button">Add button</string>
+    <string name="select_keycode">Select Keyboard Button</string>
+    <string name="select_icon">Select Icon</string>
+    <string name="tuner_other">Other</string>
+    <string name="left">Left</string>
+    <string name="right">Right</string>
+    <string name="restore">Restore active layout</string>
+    <string name="navbar_editor_power">Power</string>
+    <string name="navbar_editor_volup">Volume up</string>
+    <string name="navbar_editor_voldown">Volume down</string>
+    <string name="navbar_editor_title">Editor</string>
+    <string name="navbar_editor_summary">Configure navigaton bar</string>
+    <!-- SysUI Tuner: Adjust button width dialog title [CHAR LIMIT=60] -->
+    <string name="adjust_button_width">Adjust button width</string>
+
+    <!-- QuickSettings: InternetDialog: Summary for how many devices are connected to the hotspot [CHAR LIMIT=NONE] -->
+    <plurals name="quick_settings_internet_hotspot_summary_num_devices">
+        <item quantity="one">%d device connected</item>
+        <item quantity="other">%d devices connected</item>
+    </plurals>
+
+    <!-- DataSwitch Tile -->
+    <string name="qs_data_switch_label">Switch data card</string>
+
+    <!-- Name of the HD calling status bar icon. -->
+    <string name="status_bar_hd_calling">HD calling (VoLTE/VoNR)</string>
+
+    <!-- Name of the VoWiFi status bar icon. -->
+    <string name="status_bar_vowifi">Wi-Fi calling (VoWiFi)</string>
+
+    <!-- Name of the Roaming status bar icon. -->
+    <string name="status_bar_roaming">Roaming</string>
 </resources>
diff --git a/packages/SystemUI/res/values/lineage_dimens.xml b/packages/SystemUI/res/values/lineage_dimens.xml
index e1998ce3d3ab1..65b97cbfa8707 100644
--- a/packages/SystemUI/res/values/lineage_dimens.xml
+++ b/packages/SystemUI/res/values/lineage_dimens.xml
@@ -21,4 +21,11 @@
 
     <!-- Largest size an avatar might need to be drawn in the power menu user picker -->
     <dimen name="global_actions_avatar_size">24dp</dimen>
+
+    <!-- Bluetooth dialog -->
+    <dimen name="bluetooth_dialog_active_device_bottom_margin">4dp</dimen>
+    <dimen name="bluetooth_dialog_toggle_height">56dp</dimen>
+
+    <!-- HD calling icon -->
+    <dimen name="signal_icon_viewport_size">24dp</dimen>
 </resources>
diff --git a/packages/SystemUI/res/xml/status_bar_prefs.xml b/packages/SystemUI/res/xml/status_bar_prefs.xml
index 865f844733758..ddf066fb4798a 100644
--- a/packages/SystemUI/res/xml/status_bar_prefs.xml
+++ b/packages/SystemUI/res/xml/status_bar_prefs.xml
@@ -101,6 +101,21 @@
         android:key="airplane"
         android:title="@string/status_bar_airplane" />
 
+    <com.android.systemui.tuner.StatusBarSwitch
+        android:icon="@drawable/ic_statusbar_roaming"
+        android:key="roaming"
+        android:title="@string/status_bar_roaming" />
+
+    <com.android.systemui.tuner.StatusBarSwitch
+        android:icon="@drawable/ic_statusbar_hd_calling"
+        android:key="hd_calling"
+        android:title="@string/status_bar_hd_calling" />
+
+    <com.android.systemui.tuner.StatusBarSwitch
+        android:icon="@drawable/ic_statusbar_vowifi"
+        android:key="vowifi"
+        android:title="@string/status_bar_vowifi" />
+
     <!-- other weird signal stuff -->
 
     <com.android.systemui.tuner.StatusBarSwitch
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/StatusBarMobileView.java b/packages/SystemUI/src/com/android/systemui/statusbar/StatusBarMobileView.java
index fdad101ae0f65..9e1121d03b1f5 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/StatusBarMobileView.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/StatusBarMobileView.java
@@ -59,8 +59,9 @@ public class StatusBarMobileView extends BaseStatusBarFrameLayout implements Dar
     private View mInoutContainer;
     private ImageView mIn;
     private ImageView mOut;
-    private ImageView mMobile, mMobileType, mMobileRoaming;
     private View mMobileRoamingSpace;
+    private ImageView mMobile, mMobileType, mMobileHd, mMobileRoaming;
+    private View mMobileTypeSpace, mMobileHdSpace;
     @StatusBarIconView.VisibleState
     private int mVisibleState = STATE_HIDDEN;
     private DualToneHandler mDualToneHandler;
@@ -119,8 +120,9 @@ public class StatusBarMobileView extends BaseStatusBarFrameLayout implements Dar
         mMobileGroup = findViewById(R.id.mobile_group);
         mMobile = findViewById(R.id.mobile_signal);
         mMobileType = findViewById(R.id.mobile_type);
-        mMobileRoaming = findViewById(R.id.mobile_roaming);
-        mMobileRoamingSpace = findViewById(R.id.mobile_roaming_space);
+        mMobileHd = findViewById(R.id.mobile_hd);
+        mMobileHdSpace = findViewById(R.id.mobile_hd_space);
+        mMobileTypeSpace = findViewById(R.id.mobile_type_space);
         mIn = findViewById(R.id.mobile_in);
         mOut = findViewById(R.id.mobile_out);
         mInoutContainer = findViewById(R.id.inout_container);
@@ -128,6 +130,15 @@ public class StatusBarMobileView extends BaseStatusBarFrameLayout implements Dar
         mMobileDrawable = new SignalDrawable(getContext());
         mMobile.setImageDrawable(mMobileDrawable);
 
+        float signalSize = mContext.getResources().getDimensionPixelSize(
+                R.dimen.signal_icon_size);
+        float viewportSize = mContext.getResources().getDimensionPixelSize(
+                R.dimen.signal_icon_viewport_size);
+        LayoutParams lp = (LayoutParams) mMobileHd.getLayoutParams();
+        lp.height = Math.round((float) lp.height * (signalSize / viewportSize));
+        lp.width = Math.round((float) lp.width * (signalSize / viewportSize));
+        mMobileHd.setLayoutParams(lp);
+
         initDotView();
     }
 
@@ -176,8 +187,8 @@ public class StatusBarMobileView extends BaseStatusBarFrameLayout implements Dar
             mMobileType.setVisibility(View.GONE);
         }
         mMobile.setVisibility(mState.showTriangle ? View.VISIBLE : View.GONE);
-        mMobileRoaming.setVisibility(mState.roaming ? View.VISIBLE : View.GONE);
-        mMobileRoamingSpace.setVisibility(mState.roaming ? View.VISIBLE : View.GONE);
+        mMobileHd.setVisibility(mState.showHd ? View.VISIBLE : View.GONE);
+        mMobileHdSpace.setVisibility(mState.showHd ? View.VISIBLE : View.GONE);
         mIn.setVisibility(mState.activityIn ? View.VISIBLE : View.GONE);
         mOut.setVisibility(mState.activityOut ? View.VISIBLE : View.GONE);
         mInoutContainer.setVisibility((mState.activityIn || mState.activityOut)
@@ -208,14 +219,14 @@ public class StatusBarMobileView extends BaseStatusBarFrameLayout implements Dar
         }
 
         mMobile.setVisibility(state.showTriangle ? View.VISIBLE : View.GONE);
-        mMobileRoaming.setVisibility(state.roaming ? View.VISIBLE : View.GONE);
-        mMobileRoamingSpace.setVisibility(state.roaming ? View.VISIBLE : View.GONE);
+        mMobileHd.setVisibility(state.showHd ? View.VISIBLE : View.GONE);
+        mMobileHdSpace.setVisibility(state.showHd ? View.VISIBLE : View.GONE);
         mIn.setVisibility(state.activityIn ? View.VISIBLE : View.GONE);
         mOut.setVisibility(state.activityOut ? View.VISIBLE : View.GONE);
         mInoutContainer.setVisibility((state.activityIn || state.activityOut)
                 ? View.VISIBLE : View.GONE);
 
-        needsLayout |= state.roaming != mState.roaming
+        needsLayout |= state.showHd != mState.showHd
                 || state.activityIn != mState.activityIn
                 || state.activityOut != mState.activityOut
                 || state.showTriangle != mState.showTriangle;
@@ -233,7 +244,7 @@ public class StatusBarMobileView extends BaseStatusBarFrameLayout implements Dar
         mIn.setImageTintList(color);
         mOut.setImageTintList(color);
         mMobileType.setImageTintList(color);
-        mMobileRoaming.setImageTintList(color);
+        mMobileHd.setImageTintList(color);
         mDotView.setDecorColor(tint);
         mDotView.setIconColor(tint, false);
     }
@@ -254,7 +265,7 @@ public class StatusBarMobileView extends BaseStatusBarFrameLayout implements Dar
         mIn.setImageTintList(list);
         mOut.setImageTintList(list);
         mMobileType.setImageTintList(list);
-        mMobileRoaming.setImageTintList(list);
+        mMobileHd.setImageTintList(list);
         mDotView.setDecorColor(color);
     }
 
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/connectivity/MobileSignalController.java b/packages/SystemUI/src/com/android/systemui/statusbar/connectivity/MobileSignalController.java
index 03f75a7b72044..08170231ba4d5 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/connectivity/MobileSignalController.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/connectivity/MobileSignalController.java
@@ -15,6 +15,8 @@
  */
 package com.android.systemui.statusbar.connectivity;
 
+import static android.telephony.ims.stub.ImsRegistrationImplBase.REGISTRATION_TECH_IWLAN;
+import static android.telephony.ims.stub.ImsRegistrationImplBase.REGISTRATION_TECH_NONE;
 import static android.telephony.TelephonyManager.UNKNOWN_CARRIER_ID;
 
 import android.content.Context;
@@ -30,8 +32,16 @@ import android.telephony.SignalStrength;
 import android.telephony.SubscriptionInfo;
 import android.telephony.SubscriptionManager;
 import android.telephony.TelephonyManager;
+import android.telephony.ims.ImsException;
+import android.telephony.ims.ImsMmTelManager;
+import android.telephony.ims.ImsReasonInfo;
+import android.telephony.ims.ImsRegistrationAttributes;
+import android.telephony.ims.ImsStateCallback;
+import android.telephony.ims.RegistrationManager.RegistrationCallback;
+import android.telephony.ims.feature.MmTelFeature;
 import android.text.Html;
 import android.text.TextUtils;
+import android.util.ArraySet;
 import android.util.Log;
 
 import com.android.internal.annotations.VisibleForTesting;
@@ -43,8 +53,11 @@ import com.android.settingslib.mobile.MobileStatusTracker.MobileStatus;
 import com.android.settingslib.mobile.MobileStatusTracker.SubscriptionDefaults;
 import com.android.settingslib.mobile.TelephonyIcons;
 import com.android.settingslib.net.SignalStrengthUtil;
+import com.android.systemui.Dependency;
 import com.android.systemui.R;
+import com.android.systemui.statusbar.phone.StatusBarIconController;
 import com.android.systemui.statusbar.pipeline.mobile.util.MobileMappingsProxy;
+import com.android.systemui.tuner.TunerService;
 import com.android.systemui.util.CarrierConfigTracker;
 
 import java.io.PrintWriter;
@@ -76,12 +89,31 @@ public class MobileSignalController extends SignalController<MobileState, Mobile
     boolean mInflateSignalStrengths = false;
     @VisibleForTesting
     final MobileStatusTracker mMobileStatusTracker;
+    private final ImsMmTelManager mImsMmTelManager;
 
     // Save the previous STATUS_HISTORY_SIZE states for logging.
     private final String[] mMobileStatusHistory = new String[STATUS_HISTORY_SIZE];
     // Where to copy the next state into.
     private int mMobileStatusHistoryIndex;
 
+    private boolean mHideRoaming;
+    private boolean mHideHd;
+    private boolean mHideVowifi;
+
+    private static final String KEY_ROAMING = "roaming";
+    private static final String KEY_HD_CALLING = "hd_calling";
+    private static final String KEY_VOWIFI = "vowifi";
+
+    private final TunerService mTunerService;
+    private final TunerService.Tunable mTunable = (key, value) -> {
+        final ArraySet<String> hideList =
+                StatusBarIconController.getIconHideList(mContext, value);
+        mHideRoaming = hideList.contains(KEY_ROAMING);
+        mHideHd = hideList.contains(KEY_HD_CALLING);
+        mHideVowifi = hideList.contains(KEY_VOWIFI);
+        notifyListeners();
+    };
+
     private final MobileStatusTracker.Callback mMobileCallback =
             new MobileStatusTracker.Callback() {
                 private String mLastStatus;
@@ -112,6 +144,56 @@ public class MobileSignalController extends SignalController<MobileState, Mobile
                 }
             };
 
+    private final ImsStateCallback mImsStateCallback = new ImsStateCallback() {
+        @Override
+        public void onUnavailable(int reason) {
+            Log.d(mTag, "ImsStateCallback.onUnavailable: reason=" + reason);
+            unregisterImsListeners();
+        }
+
+        @Override
+        public void onAvailable() {
+            Log.d(mTag, "ImsStateCallback.onAvailable");
+            registerImsListeners();
+        }
+
+        @Override
+        public void onError() {
+            Log.e(mTag, "ImsStateCallback.onError");
+            unregisterImsListeners();
+        }
+    };
+
+    private final RegistrationCallback mRegistrationCallback = new RegistrationCallback() {
+        @Override
+        public void onRegistered(ImsRegistrationAttributes attributes) {
+            Log.d(mTag, "onRegistered: attributes=" + attributes);
+            mCurrentState.imsRegistered = true;
+            mCurrentState.imsRegistrationTech = attributes.getRegistrationTechnology();
+            notifyListenersIfNecessary();
+        }
+
+        @Override
+        public void onUnregistered(ImsReasonInfo info) {
+            Log.d(mTag, "onUnregistered: info=" + info);
+            mCurrentState.imsRegistered = false;
+            mCurrentState.imsRegistrationTech = REGISTRATION_TECH_NONE;
+            notifyListenersIfNecessary();
+        }
+    };
+
+    private ImsMmTelManager.CapabilityCallback mCapabilityCallback
+            = new ImsMmTelManager.CapabilityCallback() {
+        @Override
+        public void onCapabilitiesStatusChanged(MmTelFeature.MmTelCapabilities config) {
+            mCurrentState.imsVoiceCapable =
+                    config.isCapable(MmTelFeature.MmTelCapabilities.CAPABILITY_TYPE_VOICE);
+            Log.d(mTag, "onCapabilitiesStatusChanged: imsVoiceCapable="
+                    + mCurrentState.imsVoiceCapable);
+            notifyListenersIfNecessary();
+        }
+    };
+
     // TODO: Reduce number of vars passed in, if we have the NetworkController, probably don't
     // need listener lists anymore.
     public MobileSignalController(
@@ -158,6 +240,8 @@ public class MobileSignalController extends SignalController<MobileState, Mobile
             }
         };
         mMobileStatusTracker = mobileStatusTrackerFactory.createTracker(mMobileCallback);
+        mImsMmTelManager = ImsMmTelManager.createForSubscriptionId(info.getSubscriptionId());
+        mTunerService = Dependency.get(TunerService.class);
     }
 
     void setConfiguration(Config config) {
@@ -202,6 +286,13 @@ public class MobileSignalController extends SignalController<MobileState, Mobile
         mContext.getContentResolver().registerContentObserver(Global.getUriFor(
                 Global.MOBILE_DATA + mSubscriptionInfo.getSubscriptionId()),
                 true, mObserver);
+        try {
+            mImsMmTelManager.registerImsStateCallback(mContext.getMainExecutor(),
+                    mImsStateCallback);
+        } catch (ImsException e) {
+            Log.e(mTag, "failed to call registerImsStateCallback", e);
+        }
+        mTunerService.addTunable(mTunable, StatusBarIconController.ICON_HIDE_LIST);
     }
 
     /**
@@ -210,6 +301,29 @@ public class MobileSignalController extends SignalController<MobileState, Mobile
     public void unregisterListener() {
         mMobileStatusTracker.setListening(false);
         mContext.getContentResolver().unregisterContentObserver(mObserver);
+        mImsMmTelManager.unregisterImsStateCallback(mImsStateCallback);
+        unregisterImsListeners();
+        mTunerService.removeTunable(mTunable);
+    }
+
+    private void registerImsListeners() {
+        try {
+            mImsMmTelManager.registerMmTelCapabilityCallback(mContext.getMainExecutor(),
+                    mCapabilityCallback);
+            mImsMmTelManager.registerImsRegistrationCallback(mContext.getMainExecutor(),
+                    mRegistrationCallback);
+        } catch (Exception e) {
+            Log.e(mTag, "unable to register ims callbacks", e);
+        }
+    }
+
+    private void unregisterImsListeners() {
+        try {
+            mImsMmTelManager.unregisterMmTelCapabilityCallback(mCapabilityCallback);
+            mImsMmTelManager.unregisterImsRegistrationCallback(mRegistrationCallback);
+        } catch (Exception e) {
+            Log.e(mTag, "unable to unregister ims callbacks", e);
+        }
     }
 
     private void updateInflateSignalStrength() {
@@ -239,7 +353,8 @@ public class MobileSignalController extends SignalController<MobileState, Mobile
                             && mCurrentState.defaultDataOff));
             boolean noInternet = mCurrentState.inetCondition == 0;
             boolean cutOut = dataDisabled || noInternet;
-            return SignalDrawable.getState(level, getNumLevels(), cutOut);
+            boolean roaming = isRoaming() && !mHideRoaming;
+            return SignalDrawable.getState(level, getNumLevels(), cutOut, roaming);
         } else if (mCurrentState.enabled) {
             return SignalDrawable.getEmptyState(getNumLevels());
         } else {
@@ -290,6 +405,7 @@ public class MobileSignalController extends SignalController<MobileState, Mobile
                 mSubscriptionInfo.getSubscriptionId(),
                 mCurrentState.roaming,
                 sbInfo.showTriangle,
+                sbInfo.showHd,
                 mCurrentState.isDefault);
         callback.setMobileDataIndicators(mobileDataIndicators);
     }
@@ -326,8 +442,18 @@ public class MobileSignalController extends SignalController<MobileState, Mobile
         int typeIcon =
                 (showDataIconInStatusBar || mConfig.alwaysShowDataRatIcon) ? dataTypeIcon : 0;
         boolean showTriangle = mCurrentState.enabled && !mCurrentState.airplaneMode;
+        boolean showHd = mCurrentState.imsRegistered && mCurrentState.imsVoiceCapable && !mHideHd;
+
+        if (mCurrentState.imsRegistrationTech == REGISTRATION_TECH_IWLAN
+                && mCurrentState.imsVoiceCapable && !mHideVowifi) {
+            typeIcon = R.drawable.ic_vowifi;
+            statusIcon = new IconState(true,
+                    ((mCurrentState.enabled && !mCurrentState.airplaneMode) ? statusIcon.icon : -1),
+                    contentDescription);
+            showHd = false;
+        }
 
-        return new SbInfo(showTriangle, typeIcon, statusIcon);
+        return new SbInfo(showTriangle, showHd, typeIcon, statusIcon);
     }
 
     @Override
@@ -599,20 +725,21 @@ public class MobileSignalController extends SignalController<MobileState, Mobile
 
     /** Box for status bar icon info */
     private static final class SbInfo {
-        final boolean showTriangle;
+        final boolean showTriangle, showHd;
         final int ratTypeIcon;
         final IconState icon;
 
-        SbInfo(boolean show, int typeIcon, IconState iconState) {
+        SbInfo(boolean show, boolean hd, int typeIcon, IconState iconState) {
             showTriangle = show;
+            showHd = hd;
             ratTypeIcon = typeIcon;
             icon = iconState;
         }
 
         @Override
         public String toString() {
-            return "SbInfo: showTriangle=" + showTriangle + " ratTypeIcon=" + ratTypeIcon
-                    + " icon=" + icon;
+            return "SbInfo: showTriangle=" + showTriangle + " showHd=" + showHd
+                    + " ratTypeIcon=" + ratTypeIcon + " icon=" + icon;
         }
     }
 }
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/connectivity/MobileState.kt b/packages/SystemUI/src/com/android/systemui/statusbar/connectivity/MobileState.kt
index 1fb6a982fdf30..c798367e84a24 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/connectivity/MobileState.kt
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/connectivity/MobileState.kt
@@ -22,6 +22,7 @@ import android.telephony.ServiceState
 import android.telephony.SignalStrength
 import android.telephony.TelephonyDisplayInfo
 import android.telephony.TelephonyManager
+import android.telephony.ims.stub.ImsRegistrationImplBase.REGISTRATION_TECH_NONE
 import com.android.internal.annotations.VisibleForTesting
 import com.android.settingslib.SignalIcon.MobileIconGroup
 import com.android.settingslib.Utils
@@ -46,6 +47,9 @@ internal class MobileState(
     @JvmField var dataState: Int = TelephonyManager.DATA_DISCONNECTED,
     // Tracks the on/off state of the defaultDataSubscription
     @JvmField var defaultDataOff: Boolean = false,
+    @JvmField var imsRegistered: Boolean = false,
+    @JvmField var imsRegistrationTech: Int = REGISTRATION_TECH_NONE,
+    @JvmField var imsVoiceCapable: Boolean = false
 ) : ConnectivityState() {
 
     @JvmField var telephonyDisplayInfo = TelephonyDisplayInfo(TelephonyManager.NETWORK_TYPE_UNKNOWN,
@@ -95,6 +99,9 @@ internal class MobileState(
         roaming = o.roaming
         dataState = o.dataState
         defaultDataOff = o.defaultDataOff
+        imsRegistered = o.imsRegistered
+        imsRegistrationTech = o.imsRegistrationTech
+        imsVoiceCapable = o.imsVoiceCapable
 
         telephonyDisplayInfo = o.telephonyDisplayInfo
         serviceState = o.serviceState
@@ -176,6 +183,9 @@ internal class MobileState(
         builder.append("userSetup=$userSetup,")
         builder.append("dataState=$dataState,")
         builder.append("defaultDataOff=$defaultDataOff,")
+        builder.append("imsRegistered=$imsRegistered,")
+        builder.append("imsRegistrationTech=$imsRegistrationTech,")
+        builder.append("imsVoiceCapable=$imsVoiceCapable,")
 
         // Computed properties
         builder.append("showQuickSettingsRatIcon=${showQuickSettingsRatIcon()},")
@@ -203,6 +213,9 @@ internal class MobileState(
             "userSetup",
             "dataState",
             "defaultDataOff",
+            "imsRegistered",
+            "imsRegistrationTech",
+            "imsVoiceCapable",
             "showQuickSettingsRatIcon",
             "voiceServiceState",
             "isInService",
@@ -228,6 +241,9 @@ internal class MobileState(
                 userSetup,
                 dataState,
                 defaultDataOff,
+                imsRegistered,
+                imsRegistrationTech,
+                imsVoiceCapable,
                 showQuickSettingsRatIcon(),
                 getVoiceServiceState(),
                 isInService(),
@@ -259,6 +275,9 @@ internal class MobileState(
         if (roaming != other.roaming) return false
         if (dataState != other.dataState) return false
         if (defaultDataOff != other.defaultDataOff) return false
+        if (imsRegistered != other.imsRegistered) return false
+        if (imsRegistrationTech != other.imsRegistrationTech) return false
+        if (imsVoiceCapable != other.imsVoiceCapable) return false
         if (telephonyDisplayInfo != other.telephonyDisplayInfo) return false
         if (serviceState != other.serviceState) return false
         if (signalStrength != other.signalStrength) return false
@@ -281,6 +300,9 @@ internal class MobileState(
         result = 31 * result + roaming.hashCode()
         result = 31 * result + dataState
         result = 31 * result + defaultDataOff.hashCode()
+        result = 31 * result + imsRegistered.hashCode()
+        result = 31 * result + imsRegistrationTech.hashCode()
+        result = 31 * result + imsVoiceCapable.hashCode()
         result = 31 * result + telephonyDisplayInfo.hashCode()
         result = 31 * result + (serviceState?.hashCode() ?: 0)
         result = 31 * result + (signalStrength?.hashCode() ?: 0)
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/connectivity/SignalCallback.kt b/packages/SystemUI/src/com/android/systemui/statusbar/connectivity/SignalCallback.kt
index 501f694404561..1d9f96a7515af 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/connectivity/SignalCallback.kt
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/connectivity/SignalCallback.kt
@@ -154,6 +154,7 @@ data class MobileDataIndicators(
     @JvmField val subId: Int,
     @JvmField val roaming: Boolean,
     @JvmField val showTriangle: Boolean,
+    @JvmField val showHd: Boolean,
     @JvmField val isDefault: Boolean
 ) {
     override fun toString(): String {
@@ -170,6 +171,7 @@ data class MobileDataIndicators(
                 .append(",subId=").append(subId)
                 .append(",roaming=").append(roaming)
                 .append(",showTriangle=").append(showTriangle)
+                .append(",showHd=").append(showHd)
                 .append(",isDefault=").append(isDefault)
                 .append(']').toString()
     }
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/connectivity/WifiSignalController.java b/packages/SystemUI/src/com/android/systemui/statusbar/connectivity/WifiSignalController.java
index ac48e7dc46591..9d1a5535473de 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/connectivity/WifiSignalController.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/connectivity/WifiSignalController.java
@@ -149,7 +149,7 @@ public class WifiSignalController extends SignalController<WifiState, IconGroup>
                 mCurrentState.activityIn, mCurrentState.activityOut, dataContentDescription,
                 dataContentDescriptionHtml, description,
                 mCurrentState.subId, /* roaming= */ false, /* showTriangle= */ true,
-                /* isDefault= */ qsIcon != null
+                /* showHd */ false, /* isDefault= */ qsIcon != null
         );
         callback.setMobileDataIndicators(mobileDataIndicators);
     }
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/phone/StatusBarSignalPolicy.java b/packages/SystemUI/src/com/android/systemui/statusbar/phone/StatusBarSignalPolicy.java
index de7bf3c021dd0..a8cb1619d990a 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/phone/StatusBarSignalPolicy.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/phone/StatusBarSignalPolicy.java
@@ -195,8 +195,7 @@ public class StatusBarSignalPolicy implements SignalCallback,
             newState.activityIn = in;
             newState.activityOut = out;
             newState.contentDescription = indicators.statusIcon.contentDescription;
-            MobileIconState first = getFirstMobileState();
-            newState.signalSpacerVisible = first != null && first.typeId != 0;
+            updateShowWifiSignalSpacer(newState);
         }
         newState.slot = mSlotWifi;
         newState.airplaneSpacerVisible = mIsAirplaneMode;
@@ -206,7 +205,7 @@ public class StatusBarSignalPolicy implements SignalCallback,
 
     private void updateShowWifiSignalSpacer(WifiIconState state) {
         MobileIconState first = getFirstMobileState();
-        state.signalSpacerVisible = first != null && first.typeId != 0;
+        state.signalSpacerVisible = first != null && (first.typeId != 0 || first.showHd);
     }
 
     private void updateWifiIconWithState(WifiIconState state) {
@@ -257,9 +256,10 @@ public class StatusBarSignalPolicy implements SignalCallback,
             return;
         }
 
-        // Visibility of the data type indicator changed
-        boolean typeChanged = indicators.statusType != state.typeId
-                && (indicators.statusType == 0 || state.typeId == 0);
+        // Visibility of the data type indicator or HD icon changed
+        boolean typeChanged = (indicators.statusType != state.typeId
+                && (indicators.statusType == 0 || state.typeId == 0))
+                || indicators.showHd != state.showHd;
 
         state.visible = indicators.statusIcon.visible && !mHideMobile;
         state.strengthId = indicators.statusIcon.icon;
@@ -267,9 +267,12 @@ public class StatusBarSignalPolicy implements SignalCallback,
         state.contentDescription = indicators.statusIcon.contentDescription;
         state.typeContentDescription = indicators.typeContentDescription;
         state.showTriangle = indicators.showTriangle;
-        state.roaming = indicators.roaming;
+        state.showHd = indicators.showHd;
         state.activityIn = indicators.activityIn && mActivityEnabled;
         state.activityOut = indicators.activityOut && mActivityEnabled;
+        state.typeSpacerVisible = mMobileStates.size() > 1
+               && mMobileStates.get(1).subId == state.subId
+               && state.typeId != 0 && state.typeId != R.drawable.ic_vowifi;
 
         if (DEBUG) {
             Log.d(TAG, "MobileIconStates: "
@@ -566,9 +569,10 @@ public class StatusBarSignalPolicy implements SignalCallback,
         public int strengthId;
         public int typeId;
         public boolean showTriangle;
-        public boolean roaming;
+        public boolean showHd;
         public boolean needsLeadingPadding;
         public CharSequence typeContentDescription;
+        public boolean typeSpacerVisible;
 
         private MobileIconState(int subId) {
             super();
@@ -588,7 +592,7 @@ public class StatusBarSignalPolicy implements SignalCallback,
                     && strengthId == that.strengthId
                     && typeId == that.typeId
                     && showTriangle == that.showTriangle
-                    && roaming == that.roaming
+                    && showHd == that.showHd
                     && needsLeadingPadding == that.needsLeadingPadding
                     && Objects.equals(typeContentDescription, that.typeContentDescription);
         }
@@ -597,7 +601,7 @@ public class StatusBarSignalPolicy implements SignalCallback,
         public int hashCode() {
 
             return Objects
-                    .hash(super.hashCode(), subId, strengthId, typeId, showTriangle, roaming,
+                    .hash(super.hashCode(), subId, strengthId, typeId, showTriangle, showHd,
                             needsLeadingPadding, typeContentDescription);
         }
 
@@ -613,7 +617,7 @@ public class StatusBarSignalPolicy implements SignalCallback,
             other.strengthId = strengthId;
             other.typeId = typeId;
             other.showTriangle = showTriangle;
-            other.roaming = roaming;
+            other.showHd = showHd;
             other.needsLeadingPadding = needsLeadingPadding;
             other.typeContentDescription = typeContentDescription;
         }
@@ -631,8 +635,9 @@ public class StatusBarSignalPolicy implements SignalCallback,
 
         @Override public String toString() {
             return "MobileIconState(subId=" + subId + ", strengthId=" + strengthId
-                    + ", showTriangle=" + showTriangle + ", roaming=" + roaming
-                    + ", typeId=" + typeId + ", visible=" + visible + ")";
+                    + ", showTriangle=" + showTriangle + ", showHd=" + showHd
+                    + ", typeId=" + typeId + ", typeSpacerVisible=" + typeSpacerVisible
+                    + ", visible=" + visible + ")";
         }
     }
 }
-- 
2.34.1

